### PR TITLE
Fixed sql server limit & offset clause

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

File: gateway/modules/crud/sql/helpers.go [hunk 1]
```diff
@@ -248,3 +249,64 @@ func (s *SQL) processJoins(ctx context.Context, query *goqu.SelectDataset, join
...
+	dollarEndIndex := startIndex + len(replace) + 1
+	if dollarEndIndex > len(sqlString) {
+		return "", sqlString
+	}
+	tempArr := sqlString[dollarEndIndex:]
+	dollarValue := ""
+	for index, value := range tempArr {
+		dollarValue += string(value)
+		if unicode.IsSpace(value) || len(tempArr)-1 == index {
+			endIndex = dollarEndIndex + index + 1
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/helpers.go:271]: No need to do + 1 here

File: gateway/modules/crud/sql/helpers.go [hunk 2]
```diff
@@ -248,3 +249,64 @@ func (s *SQL) processJoins(ctx context.Context, query *goqu.SelectDataset, join
...
+			break
+		}
+	}
+	dollarValue = strings.TrimSpace(dollarValue)
+	arr1 := sqlString[:startIndex]
+	arr2 := sqlString[endIndex:]
+	sqlString = arr1 + replaceWith(dollarValue) + " " + arr2
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/helpers.go:278]:
```suggestion
	sqlString = arr1 + replaceWith(dollarValue) + arr2
```

File: gateway/modules/crud/sql/helpers.go [hunk 3]
```diff
@@ -248,3 +249,64 @@ func (s *SQL) processJoins(ctx context.Context, query *goqu.SelectDataset, join
...
+			return ""
+		})
+
+		_, sqlString = replaceSQLOperationWithPlaceHolder("LIMIT", sqlString, func(value string) string {
+			return fmt.Sprintf("OFFSET %s ROWS FETCH NEXT %s ROWS ONLY", offsetValue, value)
+		})
+		return sqlString
+	}
+	if req.Options.Skip != nil {
+		_, sqlString = replaceSQLOperationWithPlaceHolder("LIMIT", sqlString, func(value string) string {
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/helpers.go:294]: Is this step really required

File: gateway/modules/crud/sql/helpers.go [hunk 4]
```diff
@@ -248,3 +249,64 @@ func (s *SQL) processJoins(ctx context.Context, query *goqu.SelectDataset, join
...
+	}
+	dollarValue = strings.TrimSpace(dollarValue)
+	arr1 := sqlString[:startIndex]
+	arr2 := sqlString[endIndex:]
+	sqlString = arr1 + replaceWith(dollarValue) + " " + arr2
+	return dollarValue, strings.TrimSpace(sqlString)
+}
+
+func mutateSQLServerLimitAndOffsetOperation(sqlString string, isDefaultLimit bool, req *model.ReadRequest) string {
+	if req.Options.Skip != nil && req.Options.Limit != nil && !isDefaultLimit {
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/helpers.go:283]:
```suggestion
	if req.Options.Skip != nil && req.Options.Limit != nil {
```

Comment (YourTechBud) [inline gateway/modules/crud/sql/helpers.go:283]: No need of defaultLimit here

File: gateway/modules/crud/sql/read.go [hunk 5]
```diff
@@ -30,7 +30,8 @@ func (s *SQL) generateReadQuery(ctx context.Context, col string, req *model.Read
 	if req.Options == nil {
 		req.Options = &model.ReadOptions{}
 	}
-	if req.Options.Limit == nil {
+	isDefaultLimit := req.Options.Limit == nil
+	if isDefaultLimit {
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/read.go:34]:
```suggestion
	if req.Options.Limit == nil {
```

File: gateway/modules/crud/sql/read.go [hunk 6]
```diff
@@ -148,6 +152,10 @@ func (s *SQL) generateReadQuery(ctx context.Context, col string, req *model.Read
 
 	sqlString = strings.Replace(sqlString, "\"", "", -1)
 
+	if model.DBType(s.dbType) == model.SQLServer {
+		sqlString = mutateSQLServerLimitAndOffsetOperation(sqlString, isDefaultLimit, req)
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/read.go:156]:
```suggestion
		sqlString = mutateSQLServerLimitAndOffsetOperation(sqlString, req)
```

File: gateway/modules/crud/sql/read.go [hunk 7]
```diff
@@ -64,6 +65,9 @@ func (s *SQL) generateReadQuery(ctx context.Context, col string, req *model.Read
 			}
 		}
 		if req.Options.Skip != nil {
+			if model.DBType(s.dbType) == model.SQLServer && len(req.Options.Sort) == 0 {
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/read.go:68]: Let the database throw an error

**Review by YourTechBud — CHANGES_REQUESTED** [review]
Really loved the replace logic! Amazing!!

File: gateway/modules/crud/sql/helpers.go (hunk 3 above)
Comment (sharadregoti) [reply gateway/modules/crud/sql/helpers.go:294]: Not required

File: gateway/modules/crud/sql/read_test.go [hunk 8]
```diff
@@ -532,12 +532,24 @@ func TestSQL_generateReadQuery(t *testing.T) {
 			args: args{project: "test", col: "table",
 				req: &model.ReadRequest{
 					Find:      map[string]interface{}{"Column1": map[string]interface{}{"$eq": 1}},
-					Options:   &model.ReadOptions{Skip: iti(2), Limit: iti(10)},
+					Options:   &model.ReadOptions{Skip: iti(2), Limit: iti(10), Sort: []string{"age"}},
 					Operation: "all"}},
-			want:    []string{"SELECT * FROM test.table WHERE (Column1 = @p1) LIMIT @p2 OFFSET @p3"},
+			want:    []string{"SELECT * FROM test.table WHERE (Column1 = @p1) ORDER BY age ASC OFFSET @p3 ROWS FETCH NEXT @p2 ROWS ONLY"},
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/read_test.go:537]: This test case is wrong. Check if skip and limit actually work in sqlserver. Give different values for skip and limit and manually confirm results

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: gateway/modules/crud/sql/read_test.go (hunk 8 above)
Comment (sharadregoti) [reply gateway/modules/crud/sql/read_test.go:537]: What's wrong in the test case.
Do you want me to change the integration tests?
I did test it manually on SQL server & it worked

Comment (YourTechBud) [reply gateway/modules/crud/sql/read_test.go:537]: I think the skip and limit numbers will get replaced here. Confirm please

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**